### PR TITLE
use apiservice.status to break apart controller and handling concerns

### PR DESCRIPTION
**Pull request by deads2k**
Still needs tests.

This starts breaking the handler and controller aspects of the aggregator by making use of status and conditions instead of actually running a specific check on demand.

@kubernetes/sig-api-machinery-pr-reviews 
@luxas since you've been asking

### CODE REVIEW

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -140,46 +135,51 @@ func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.Deleg
 		return nil, err
 	}
 
+	apiregistrationClient := internalclientset.NewForConfigOrDie(c.Config.GenericConfig.LoopbackClientConfig)
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: We have an error return type here. No need for the ...OrDie variant.

File: staging/src/k8s.io/kube-aggregator/pkg/registry/apiservice/etcd/etcd.go [hunk 2]
```diff
@@ -50,3 +51,28 @@ func NewREST(scheme *runtime.Scheme, optsGetter generic.RESTOptionsGetter) *REST
...
+}
+
+type StatusREST struct {
+	store *genericregistry.Store
+}
+
+var _ = rest.Updater(&StatusREST{})
+
+func (r *StatusREST) New() runtime.Object {
+	return &extensions.Deployment{}
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/registry/apiservice/etcd/etcd.go]: extensions?

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+	inAPIService, err := c.apiServiceLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	apiService := &apiregistration.APIService{}
+	if err := apiregistration.DeepCopy_apiregistration_APIService(inAPIService, apiService, cloner); err != nil {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: a nil cloner should work here too. You don't have embedded runtime.Objects where you need that.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+		availableCondition.Message = "Local APIServices are always available"
+		apiregistration.SetAPIServiceCondition(apiService, availableCondition)
+		_, err := c.apiServiceClient.APIServices().UpdateStatus(apiService)
+		return err
+	}
+
+	service, err := c.serviceLister.Services(apiService.Spec.Service.Namespace).Get(apiService.Spec.Service.Name)
+	if apierrors.IsNotFound(err) {
+		availableCondition.Status = apiregistration.ConditionFalse
+		availableCondition.Reason = "NoService"
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: ServiceNotFound would be clearer.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+		availableCondition.Message = fmt.Sprintf("service/%s in %q is not present", apiService.Spec.Service.Name, apiService.Spec.Service.Namespace)
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: service "%s/%s" would be clearer

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+		apiregistration.SetAPIServiceCondition(apiService, availableCondition)
+		_, err := c.apiServiceClient.APIServices().UpdateStatus(apiService)
+		return err
+	} else if err != nil {
+		availableCondition.Status = apiregistration.ConditionUnknown
+		availableCondition.Reason = "MissingServiceConfirmation"
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: I wouldn't know what `MissingServiceConfirmation` means, sounds confusing.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncFn(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: nit: no space

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+			return
+		}
+	}
+	for _, apiService := range c.getAPIServicesFor(castObj) {
+		c.enqueue(apiService)
+	}
+}
+
+func (c *AvailableConditionController) addEndpoints(obj interface{}) {
+	fmt.Printf("creating endpoints %v \n", obj.(*v1.Endpoints).Name)
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: debug output?

File: staging/src/k8s.io/kube-aggregator/pkg/registry/apiservice/strategy.go [hunk 9]
```diff
@@ -74,6 +75,44 @@ func (apiServerStrategy) ValidateUpdate(ctx genericapirequest.Context, obj, old
...
+
+func (apiServerStatusStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, old runtime.Object) {
+	newAPIService := obj.(*apiregistration.APIService)
+	oldAPIService := old.(*apiregistration.APIService)
+	newAPIService.Spec = oldAPIService.Spec
+	newAPIService.Labels = oldAPIService.Labels
+	newAPIService.Annotations = oldAPIService.Annotations
+	newAPIService.Finalizers = oldAPIService.Finalizers
+	newAPIService.OwnerReferences = oldAPIService.OwnerReferences
+}
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/registry/apiservice/strategy.go]: this list will be invalid when anything is added to ObjectMeta. Do we have a more elegant pattern to do that elsewhere?

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: > a nil cloner should work here too. You don't have embedded runtime.Objects where you need that.

I tried nil.  It panic'ed.  I didn't dig into why.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go (hunk 5 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: > service "%s/%s" is not present would be clearer

That doesn't match CLI patterns for namespaces.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go (hunk 6 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: > I wouldn't know what MissingServiceConfirmation means, sounds confusing.

Any alternative suggestions?

Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: Just `ServiceError` or `ServiceAccessError`?

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go (hunk 5 above)
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go]: true. Never liked the CLI pattern ;-) If it matches what we have there, it's fine.

File: staging/src/k8s.io/kube-aggregator/pkg/registry/apiservice/strategy.go (hunk 9 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/registry/apiservice/strategy.go]: > this list will be invalid when anything is added to ObjectMeta. Do we have a more elegant pattern to do that elsewhere?

Not to my knowledge.  Other resources have historically leaked and allowed annotation and label updates via status.  I'll open an issue for API machinery.

Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/registry/apiservice/strategy.go]: Opened https://github.com/kubernetes/kubernetes/issues/45539

**Comment by deads2k** [comment]
comments addressed.

**Pull request merged**